File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        while root:
            if p.val < root.val and q.val < root.val:
                root = root.left          # 둘 다 왼쪽 → 왼쪽으로
            elif p.val > root.val and q.val > root.val:
                root = root.right         # 둘 다 오른쪽 → 오른쪽으로
            else:
                return root               # 갈라짐 (또는 하나가 나 자신) = LCA

File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    bool helper(TreeNode* root, int min, int max) {
        if (root == nullptr) {
            return true;
        }

        if (root -> val >= max) {
            return false;
        }

        if (root -> val <= min) {
            return false;
        }

        return helper(root -> left, min, root -> val) && helper(root -> right, root -> val, max);
    }

public:
    bool isValidBST(TreeNode* root) {
        if (root == nullptr) {
            return true;
        }

        return helper(root -> left, INT_MIN, root -> val) && helper(root -> right, root -> val, INT_MAX);
    }
};
